machine.go

```diff
@@ -226,22 +237,35 @@ func NewMachine(ctx context.Context, cfg Config, opts ...Opt) (*Machine, error)
 // Start will iterate through the handler list and call each handler. If an
 // error occurred during handler execution, that error will be returned. If the
 // handlers succeed, then this will start the VMM instance.
+// Start may only be called once per Machine.  Subsequent calls will return
+// ErrAlreadyStarted.
 func (m *Machine) Start(ctx context.Context) error {
 	m.logger.Debug("Called Machine.Start()")
+	alreadyStarted := true
+	m.startOnce.Do(func() {
```

Comment:
@nmeyerhans and I were talking about this yesterday and we're both leaning toward enforcing the limit here.  Fundamentally, a Firecracker microVM is a one-shot thing; each microVM is tied to exactly one VMM process, and then when the VMM process exits the microVM is gone.  There isn't a notion of restarting a microVM in Firecracker; the closest that you can get is to reuse the same settings and the same disk image, and then launch a new one.

I think it makes sense to tie the `Machine` object to represent exactly one microVM, and to have this same restriction.  As a convenience, we might want to add a `Copy` or `Clone` method that spits out another independent `Machine` if you do want to "restart" it in the future.

Reply:
I'm really torn here. I think some things make sense and I get since you can't start an instance multiple times it might make sense to reflect that in code. However, there is a distinction in that the `Machine` will not start an instance multiple times but start the whole life cycle multiple times. I wonder if it makes sense to leave the control up to the user. If they want to go through the `Machine`'s start and stop in cycles, they can. I am not too sure which approach is better, but I am leaning more towards letting users control that. Thoughts?

Reply:
I think it's an easier programming model (both for us as authors of this package and for users) to assert that a `Machine`'s lifecycle is one-way.  It makes it clear that each instance is independent and cannot be restarted, and makes it easier to reason about (i.e., you get guarantees that you can program against).  For users who want to go in cycles, I think a `Copy` or `Clone` convenience method makes sense.

Reply:
It definitely is an easier programming model for us, but I don't see it as being any more easier on the user at least. However, if the `Machine` object ever does get complex, it'll make that easier to reason about. I just don't know if it'll ever get that complex. Meh, I'm okay with it lol.